ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveRewriteToDataSketchesRules.java

```diff
@@ -368,4 +389,216 @@ void rewrite(AggregateCall aggCall) {
+        relBuilder.project(projExprs);
+        ImmutableBitSet groupSets = ImmutableBitSet.range(partitionKeys.size());
+
+        SqlAggFunction aggFunction = (SqlAggFunction) getSqlOperator(DataSketchesFunctions.DATA_TO_SKETCH);
+        boolean distinct = false;
+        boolean approximate = true;
+        boolean ignoreNulls = true;
+        List<Integer> argList = Lists.newArrayList(partitionKeys.size());
+        int filterArg = -1;
+        RelCollation collation = RelCollations.EMPTY;
+        RelDataType type = rexBuilder.deriveReturnType(aggFunction, Collections.emptyList());
+        String name = aggFunction.getName();
+        AggregateCall newAgg = AggregateCall.create(aggFunction, distinct, approximate, ignoreNulls, argList, filterArg,
+                      collation, type, name);
+
+        RelNode agg = HiveRelFactories.HIVE_AGGREGATE_FACTORY.createAggregate(
+            relBuilder.build(),
+            groupSets, ImmutableList.of(groupSets),
+            Lists.newArrayList(newAgg));
+        relBuilder.push(agg);
+
+        List<RexNode> joinConditions;
+        joinConditions = Ord.zip(partitionKeys).stream().map(o -> {
+          RexNode f = relBuilder.field(2, 1, o.i);
+          return rexBuilder.makeCall(SqlStdOperatorTable.EQUALS, o.e, f);
+        }).collect(Collectors.toList());
+        relBuilder.join(JoinRelType.INNER, joinConditions);
+
+        int sketchFieldIndex = relBuilder.peek().getRowType().getFieldCount() - 1;
+        RexInputRef sketchInputRef = relBuilder.field(sketchFieldIndex);
+        SqlOperator projectOperator = getSqlOperator(DataSketchesFunctions.GET_CDF);
+
+        // NULLs will be replaced by this value - to be before / after the other values
+        // note: the sketch will ignore NULLs entirely but they will be placed at 0.0 or 1.0
+        final RexNode nullReplacement =
+            relBuilder.literal(orderKey.getNullDirection() == NullDirection.FIRST ? Float.MAX_VALUE : -Float.MAX_VALUE);
+
+        // long story short: CAST(1.0f-CDF(CAST(COALESCE(-X, nullReplacement) AS FLOAT))[0] AS targetType)
+        RexNode projRex = key;
+        projRex = rexBuilder.makeCall(SqlStdOperatorTable.COALESCE, key, nullReplacement);
+        projRex = rexBuilder.makeCast(getFloatType(), projRex);
+        projRex = rexBuilder.makeCall(projectOperator, ImmutableList.of(sketchInputRef, projRex));
+        projRex = makeItemCall(projRex, relBuilder.literal(0));
+        projRex = rexBuilder.makeCall(SqlStdOperatorTable.MINUS, relBuilder.literal(1.0f), projRex);
+        projRex = rexBuilder.makeCast(over.getType(), projRex);
+
+        return projRex;
+      }
+
+      private RexNode makeItemCall(RexNode arr, RexNode offset) {
+
+        if(getClass().desiredAssertionStatus()) {
+          try {
+            SqlKind.class.getField("ITEM");
+            throw new RuntimeException("bind SqlKind.ITEM instead of this workaround - C1.23 a02155a70a");
```

Comment:
??

Reply:
`SqlKind.ITEM` was introduced only recently in 1.23...this check is here to throw an error after we've upgraded
all these things in this method would be unneccessary if SqlKind.ITEM would be available...
a simple `rexBuilder.makeCall(SqlStdOperatorTable.ITEM, arr, offset)`